Add feature to test all kata setup.sh scripts

JKrag opened the pull request:
This PR includes addition of a test-al.sh script that test-runs the setup.sh of ALL katas, not just the basic-commit one.

To test all katas, run 
```
./test-all.sh
```
from the root folder.

It also includes a `./utils/test/test_setup.sh` can be used to test a single kata (e.g. during development)
Run it from the kata folder as:

```
../utils/test/test_setup.sh
```

While doing this, I noticed that the existing `test.sh` script currently used by the pipeline doesn't actually fail at all if there are problems. So this has now been modified to use the new test_setup.sh script and now fails as it should. (but `test.sh` still only tests the single basic-commits setup).

*NOTE:* The obvious omission in this PR is that the pipeline has not been changed to run the `test-all.sh` script as it obviously should. The reason for this is that I still need to figure out how to get `test-all.sh` to "collect" a summary and exit non-zero if there was one or more tests failing. But I figured the rest was valuable enough to get added now as is. 

Initial work on this PR was done by me, but I want to give a shout-out to @zanderhavgaard for helping me debug and fix the error handling to get it working correctly.

atombrella commented on test-all.sh:
```diff
@@ -0,0 +1,4 @@
+#!/usr/bin/env bash
+
+echo "Testing all kata setup scripts"
+find -s . -name "setup.sh" -type f -execdir ../utils/test/test_setup.sh  \;
```
`-s` is not recognized on my setup. When I delete it, I get a list of `OK` for the katas. Can you post a snippet from the man-page on your machine of what `-s` is supposed to do?

JKrag replied:
```
-s      Cause find to traverse the file hierarchies in lexicographical order, i.e., alphabetical order within each directory.  Note: `find
             -s' and `find | sort' may give different results.
```

JKrag commented:
@atombrella Thanks for the function rewrite. This was exactly what I was hoping for, and works a charm (tested on both `zsh` and `bash` on Mac so far. 
Apart from a tiny commit with some ShellCheck linting improvements, I also felt inspired and quickly added some counters and reporting of total number of failed/completed tests. Maybe a bit crude, but it felt like it was needed for "completeness" :-)

JKrag commented:
I have now added the running of test-all.sh to the pipeline, as was the original intention.
Unfortunately this fails on Ubuntu because of the default branch warning. I have already addressed this issue in another branch, so I guess I will have to add that change to here to make it work. It does muddy the water a bit, but the change was destined to be added immediately after this one anyway.

JKrag commented:
I added my default branch Handling code, and now it just fails instead on missing user.name & user.email, but still only on Ubuntu. Lets try to fix that as well.

JKrag commented:
Finally 🟢  with all tests.
I will have to address the issue on running test-all.sh in ZSH later. Simply changing the Dockerfile to ./test-all.sh is not enough, as that would still execute the content with bash as per the shebang line.

atombrella commented on .github/workflows/test.yml:
```diff
@@ -16,8 +16,10 @@ jobs:
-        ./test.sh
+        git config --global user.name "GitHub Actions"
+        git config --global user.email "nobody@localhost"
+        chmod +x ./test-all.sh
```
The exec-bit is set in the repository. Is it really necessary?

zanderhavgaard replied:
would `bash test.sh` be more precise then?

JKrag replied:
I didn't actually add the `chmod +x` line, as it was already there. (https://github.com/eficode-academy/git-katas/pull/299/files/95d4a5eacc0f58dc1df707ffe40d84881fb5edab)
I was aware that the bit is already set in the repo, but I assumed that the `chmod +x` had been added at some point "for safety", and it does make sure that the pipeline still works even if someone on e.g. windows manages to commit without the executable bit.
I guess it doesn't hurt to have this, so I see no reason not to leave it in there.

@zanderhavgaard I have no strong opinions, but I believe that the `./test-all.sh` form, running with the shell defined in the shebang in the file, is the most idiomatic way of running scripts, as it relies on the shel that the script author has declared the script to work with (and by extension also the language spec that shellcheck will lint the script against if you use that).

atombrella reviewed (approved):
Except for a small comment about "+x", LGTM.

zanderhavgaard reviewed (approved):
Looks good, ran the tests on my machine from zsh, invoking them with `./test-all.sh` and `bash test-all.sh` both works fine :+1: Tried making one of the tests fail, which produced the desired "test-failed" output, so good work :-)
